api/src/OpenTelemetry/Internal/Logging/Types.hs

```diff
@@ -13,33 +14,68 @@ module OpenTelemetry.Internal.Logging.Types (
   severityInt,
 ) where
 
+import Control.Concurrent.Async
 import qualified Data.HashMap.Strict as H
 import Data.IORef (IORef)
 import Data.Int (Int64)
 import Data.Text (Text)
+import Data.Vector (Vector)
 import OpenTelemetry.Common (Timestamp, TraceFlags)
 import OpenTelemetry.Context.Types (Context)
-import OpenTelemetry.Internal.Common.Types (InstrumentationLibrary)
+import OpenTelemetry.Internal.Common.Types (InstrumentationLibrary, ShutdownResult)
 import OpenTelemetry.Internal.Trace.Id (SpanId, TraceId)
 import OpenTelemetry.LogAttributes
 import OpenTelemetry.Resource (MaterializedResources)
 
 
+data LogProcessor body = LogProcessor
+  { logProcessorOnEmit :: IORef (ImmutableLogRecord body) -> Context -> IO ()
```

Comment:
Similarly to the other PR, we should probably account for this naming style and provide a ReadWriteLogRecord here: https://opentelemetry.io/docs/specs/otel/logs/sdk/#additional-logrecord-interfaces

Reply:
I think that is a great idea, and I wrote a typeclass for `ReadWriteLogRecord` but it is difficult to constrain this function because it is part of a record/datatype. Writing 
```haskell
logRecordProcessorOnEmit :: (ReadWriteLogRecord r) => r body -> Context -> IO ()
``` 
does not work because `r` is not in scope. using GADTs, I could write
```haskell
data LogRecordProcessor body where
  LogRecordProcessor :: (ReadWriteLogRecord r) 
    => { logRecordProcessorOnEmit :: r body -> Context -> IO ()
          , ..
          } 
    -> LogProcessor body
```
but creating a `LogRecordProcessor` gives the error that `r` is an ambiguous type variable because it could be any possible `ReadWriteLogRecord`. I'm not sure of how to deal with this.

@iand675 @andrewsmith